File: sorts/java/MergeSort.java
package sorts;

import java.io.*;

/**
 *
 * @author Raghav
 */
public class MergeSort {
    public static int[] a;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        System.out.print("Enter length : ");
        int n = Integer.parseInt(br.readLine());
        a = new int[n];
        int i;
        System.out.println("Enter array elements :");
        for(i=0;i<n;i++){
            System.out.print("Enter element " + (i+1) + " : ");
            a[i] = Integer.parseInt(br.readLine());
        }
        mergeSort(a, 0, n-1);
        System.out.println("\nSortd Array :- ");
        for(i=0;i<n;i++)
            System.out.println(a[i]);
    }

    private static void mergeSort(int[] a, int l, int u) {
        if(l<u){
            int mid = (l+u)/2;
            mergeSort(a, l, mid);
            mergeSort(a, mid+1, u);
            merge(a, l, mid, u);
        }
    }

    private static void merge(int[] a, int l, int mid, int u) {
        int[] b =new int[u-l+1];
        int i=l;
        int j=mid+1;
        int k=0;
        while(i<=mid && j<=u){
            if (a[i]<=a[j])
                b[k++] = a[i++];
            else
                b[k++] = a[j++];
        }
        while(i<=mid)
            b[k++] = a[i++];
        while(j<=u)
            b[k++] = a[j++];
        
        i=l;
        for(k=0;k<u-l+1;k++){
            a[i++]=b[k];
        }
        
    }
    
}

File: sorts/python/merge_sort.py
# Merge Sort

arr = []
n = int(input("Enter length of array "))
for k in range(0, n):
    x = int(input("Enter element "))
    arr.append(x)

mergesort(arr, 0, n-1)
print(arr)

def mergesort(arr, l, u):
    if l<u:
        mid = (int) (l+u)/2
        mergesort(arr, l, mid)
        mergesort(arr, mid + 1, u)
        merge(arr, l, mid, u)

def merge(arr, l, mid, u):
    b =  []
    i = l
    j = mid + 1
    k = 0
    while i <= mid and j <= u:
        if arr[i] <= arr[j]:
            b[k] = arr[i]
            k = k + 1
            i = i + 1
        else:
            b[k] = arr[j]
            k = k + 1
            j = j + 1
    while i <= mid:
        b[k] = arr[i]
        k = k + 1
        i = i + 1
    while j <= u:
        b[k] = arr[j]
        k = k + 1
        j = j + 1

    i = l
    for k in range(0, u-l+1):
        arr[i] = b[k]
        i = i + 1